### ISSUE TITLE
example of trainMode=2?

### ISSUE THREAD

**Issue by viksit** [facebookresearch/Starspace#39]
Could you give an example of what you mean by mapping sentence to documents?

**Comment by ncammarata** [comment]
It's not a code example, but in the paper they go into this a bit in the `Wikipedia Article Search & Sentence Matching` section

**Comment by ncammarata** [comment]
I also would be grateful if someone could provide an example. It's the closest to my personal need.

**Comment by jaseweston** [comment]
Yes exactly, this is used in the paper (https://arxiv.org/pdf/1709.03856.pdf)  Wikipedia Article Search (Task 1) on page 5 and Table 5.  But we can provide a code example, we'll get on that..

**Comment by viksit** [comment]
@jaseweston Right, let me ask a more direct question. I presume that trainMode 2 uses the labelDoc format. So lets assume you wanted to map a sentence to a number of documents.

```
this is a sentence <tab> this is a document 1 <tab> this is a document 2 ..
```

- Now, what should the "-label" parameter here be to train?

- Should the entities between the tab be the whole document, or should it be the sentences of the document? And if the latter, how do you distinguish between sentences from doc1 vs 2?

```
this is a sentence <tab> this is a sentence from document 1 <tab> this is another sentence from document 1 <tab> this is a sent from document 2 ..
```
 
and lastly, is there a "predict" function available via the CLI? Or do you guys only use this to build embeddings, that are used elsewhere?

**Comment by jaseweston** [comment]
- yes, what you want should use the labelDoc format;
- you don't need to set the label parameter, the document is the 'whole label' and it consists of its words;
- HOWEVER it seems you want to map a sentence  to a number of documents, so trainMode 2 isn't actually what you want. As that says "At training time, one label from the collection is randomly picked as the input, and the rest of the labels in the collection become the label.". But I think you want the sentence (that is, the first item, before the first tab) to be the input always, and a random choice of another item to be the target/label for that SGD step.

So I think what you actually want is trainMode 0. That selects the first thing as the input, and a random other thing as the label. I will make that clearer in the doc right now.

**Comment by jaseweston** [comment]
ok, I've tried to make that a bit clearer in the doc

**Comment by jaseweston** [comment]
Oh, and for prediction, one can do  "starspace test"  with your test set, and specify "predictionFile" and "K" to produce the top K matches out of a set of things to rank (specified by "basedoc"). See the command line arguments in the README. There are no internal DocIDs, just the actual features of the labels. 

If you wanted to use your trained model in an application likely you'd want to modify that prediction function to get that result and do something special with it... but the base code is there.

**Comment by viksit** [comment]
@jaseweston thanks, thats helpful. Another q - how do you recommend dealing with OOV words? For instance, is there a way to use the fasttext/glove dataset here?

**Comment by jaseweston** [comment]
We don't have a direct way to deal with OOV (yet!) sorry.. depending on what you mean.
You can certainly pretrain on a bigger unlabeled dataset (e.g. Wikipedia ...) and then carry on training on a smaller one, there is a flag for that for loading existing embeddings (initModel).

If there is no dataset that contains these OOV words I guess you could also split your words into parts with spaces in your original data, e.g.:
"hello there" -> "hel elo llo the her ere", but definitely not ideal.  We could think about addressing this in the future though.

I'm going to close this issue for now.

**Comment by loretoparisi** [comment]
Hey @jaseweston I think your explanation for **trainMode=1/labelDoc** was super helpful when you say `you don't need to set the label parameter, the document is the 'whole label' and it consists of its words;`, then the algorithm will randomly pick up a label from it. This was not clear in the docs before!

That said, now looking at the examples, I can see that in `wikipedia_article_search.sh` that uses **trainMode=2** the document `wikipedia250k` looks like:

```
# head -n 2 /root/starspace/data/wikipedia_train250k.txt
Imidazoleacetate 4-monooxygenase	In enzymology , an imidazoleacetate 4-monooxygenase ( ) is an enzyme that catalyzes the chemical reaction	The 4 substrates of this enzyme are 4-imidazoleacetate , NADH , H , and O , whereas its 3 products are 5-hydroxy-4-imidazoleacetate , NAD , and HO .	This enzyme belongs to the family of oxidoreductases , specifically those acting on paired donors , with O2 as oxidant and incorporation or reduction of oxygen .	The oxygen incorporated need not be derived from O2 with NADH or NADPH as one donor , and incorporation of one atom o oxygen into the other donor .	The systematic name of this enzyme class is 4-imidazoleacetate , NADH : oxygen oxidoreductase ( 5-hydroxylating ) .	Other names in common use include imidazoleacetic hydroxylase , imidazoleacetate hydroxylase , and imidazoleacetic monooxygenase .	This enzyme participates in histidine metabolism .	It employs one cofactor , FAD .
Abel Erasmus Pass	Abel Erasmus Pass is situated in the Limpopo Province , on the R36 road between Ohrigstad and Hoedspruit/Tzaneen , South Africa .	It navigates the Manoutsa section of the Limpopo Drakensberg .	Since 1959 the road passes through the 133 m long JG Strijdom Tunnel , named after old Prime Minister JG Strijdom .	The pass is 11 km long and generally follows a coach route of the late 19th century .
```

so we have

```
trainMode = 2:
Each example contains a collection of labels. At training time, one label from the collection is randomly picked as the input, and the rest of the labels in the collection become the label.
```

So it has two columns with labels and text. This document is used in `wikipedia_sentence_matching.sh` for **trainMode=3** in the same way where:

```
trainMode = 3:
Each example contains a collection of labels. At training time, two labels from the collection are randomly picked as the input and label.
```

So I don't understand why in these examples I have the `labels\tdocument` format.

**Comment by jaseweston** [comment]
On Thu, Sep 28, 2017 at 4:53 AM, Loreto Parisi <notifications@github.com>
wrote:

> Hey @jaseweston <https://github.com/jaseweston> I think your explanation
> for *trainMode=1/labelDoc* was super helpful when you say you don't need
> to set the label parameter, the document is the 'whole label' and it
> consists of its words;, then the algorithm will randomly pick up a label
> from it. This was not clear in the docs before!
>
> That said, now looking at the examples, I can see that in
> wikipedia_article_search.sh that uses *trainMode=2* the document
> wikipedia250k looks like:
>
> # head -n 2 /root/starspace/data/wikipedia_train250k.txt
> Imidazoleacetate 4-monooxygenase	In enzymology , an imidazoleacetate 4-monooxygenase ( ) is an enzyme that catalyzes the chemical reaction	The 4 substrates of this enzyme are 4-imidazoleacetate , NADH , H , and O , whereas its 3 products are 5-hydroxy-4-imidazoleacetate , NAD , and HO .	This enzyme belongs to the family of oxidoreductases , specifically those acting on paired donors , with O2 as oxidant and incorporation or reduction of oxygen .	The oxygen incorporated need not be derived from O2 with NADH or NADPH as one donor , and incorporation of one atom o oxygen into the other donor .	The systematic name of this enzyme class is 4-imidazoleacetate , NADH : oxygen oxidoreductase ( 5-hydroxylating ) .	Other names in common use include imidazoleacetic hydroxylase , imidazoleacetate hydroxylase , and imidazoleacetic monooxygenase .	This enzyme participates in histidine metabolism .	It employs one cofactor , FAD .
> Abel Erasmus Pass	Abel Erasmus Pass is situated in the Limpopo Province , on the R36 road between Ohrigstad and Hoedspruit/Tzaneen , South Africa .	It navigates the Manoutsa section of the Limpopo Drakensberg .	Since 1959 the road passes through the 133 m long JG Strijdom Tunnel , named after old Prime Minister JG Strijdom .	The pass is 11 km long and generally follows a coach route of the late 19th century .
>
> so we have
>
> trainMode = 2:
> Each example contains a collection of labels. At training time, one label from the collection is randomly picked as the input, and the rest of the labels in the collection become the label.
>
> So it has two columns with labels and text. This document is used in
> wikipedia_sentence_matching.sh for *trainMode=3* in the same way where:
>

No, not two columns... it has as many columns (if that is what you mean by
tabbed entries) as their are sentences. Not two.  This is what we call the
"collection of labels" (each label is separated by a tab)
